sw/device/lib/testing/usb_testutils_simpleserial.c

```diff
@@ -24,53 +24,65 @@ static void ss_rx(void *ssctx_v, dif_usbdev_rx_packet_info_t packet_info,
   }
 }
 
+static void ss_tx_done(void *ssctx_v, usb_testutils_xfr_result_t result) {
+  usb_testutils_ss_ctx_t *ssctx = (usb_testutils_ss_ctx_t *)ssctx_v;
+
+  ssctx->sending = false;
+}
+
 // Called periodically by the main loop to ensure characters don't
 // stick around too long
 static void ss_flush(void *ssctx_v) {
   usb_testutils_ss_ctx_t *ssctx = (usb_testutils_ss_ctx_t *)ssctx_v;
   usb_testutils_ctx_t *ctx = ssctx->ctx;
-  if (ssctx->cur_cpos <= 0) {
+  if (ssctx->cur_cpos <= 0 || ssctx->sending) {
     return;
   }
   if ((ssctx->cur_cpos & 0x3) != 0) {
     size_t bytes_written;
-    CHECK_DIF_OK(dif_usbdev_buffer_write(ctx->dev, &ssctx->cur_buf,
-                                         ssctx->chold.data_b, /*src_len=*/4,
-                                         &bytes_written));
+    CHECK_DIF_OK(
+        dif_usbdev_buffer_write(ctx->dev, &ssctx->cur_buf, ssctx->chold.data_b,
+                                ssctx->cur_cpos & 0x3, &bytes_written));
   }
   CHECK_DIF_OK(dif_usbdev_send(ctx->dev, ssctx->ep, &ssctx->cur_buf));
+  ssctx->sending = true;
   ssctx->cur_cpos = -1;  // given it to the hardware
 }
 
 // Simple send byte will gather data for a while and send
-void usb_testutils_simpleserial_send_byte(usb_testutils_ss_ctx_t *ssctx,
+bool usb_testutils_simpleserial_send_byte(usb_testutils_ss_ctx_t *ssctx,
                                           uint8_t c) {
   usb_testutils_ctx_t *ctx = ssctx->ctx;
   if (ssctx->cur_cpos == -1) {
     CHECK_DIF_OK(
         dif_usbdev_buffer_request(ctx->dev, ctx->buffer_pool, &ssctx->cur_buf));
     ssctx->cur_cpos = 0;
+  } else if (ssctx->cur_cpos >= MAX_GATHER && ssctx->sending) {
```

Comment:
That's right. It's a small change that allows us to send longer messages, and return a success/failure indication to the caller. It can respect/ignore the result code depending upon whether it wants reliable transfer or dropping.